pkg/broker/instance_operations.go

```diff
@@ -37,21 +40,78 @@ func (b Broker) Provision(ctx context.Context, instanceID string, details broker
 		return
 	}
 
-	// Create a new Atlas cluster from the generated definition
-	resultingCluster, err := b.atlas.CreateCluster(*cluster)
+	// If cluster doesn't exist create it, otherwise compare them
+	resultingCluster, err := b.atlas.GetCluster(cluster.Name)
 	if err != nil {
-		b.logger.Errorw("Failed to create Atlas cluster", "error", err, "cluster", cluster)
-		err = atlasToAPIError(err)
-		return
+		// Create a new Atlas cluster from the generated definition
+		resultingCluster, err = b.atlas.CreateCluster(*cluster)
+		if err != nil {
+			b.logger.Errorw("Failed to create Atlas cluster", "error", err, "cluster", cluster)
+			err = atlasToAPIError(err)
+			return
+		}
+
+		b.logger.Infow("Successfully started Atlas creation process", "instance_id", instanceID, "cluster", resultingCluster)
+
+		return brokerapi.ProvisionedServiceSpec{
+			IsAsync:       true,
+			OperationData: OperationProvision,
+			DashboardURL:  b.atlas.GetDashboardURL(resultingCluster.Name),
+		}, nil
 	}
 
-	b.logger.Infow("Successfully started Atlas creation process", "instance_id", instanceID, "cluster", resultingCluster)
+	err = CompareAndReturnAppropiateResponseCode(resultingCluster, cluster)
+	return
+}
 
-	return brokerapi.ProvisionedServiceSpec{
-		IsAsync:       true,
-		OperationData: OperationProvision,
-		DashboardURL:  b.atlas.GetDashboardURL(resultingCluster.Name),
-	}, nil
+// CompareAndReturnAppropiateResponseCode converts structs to maps and afterwards returns the appropiate response code
+func CompareAndReturnAppropiateResponseCode(resultingCluster *atlas.Cluster, cluster *atlas.Cluster) error {
```

Comment:
I don't really understand. Can't we just compare the two deserialised cluster objects to see if the structs are equal?

Reply:
I tried using reflect.DeepEqual but it gave incorrect results.